masonry/src/widget/text_area.rs

```diff
@@ -709,7 +713,50 @@ impl<const EDITABLE: bool> Widget for TextArea<EDITABLE> {
             TextEvent::FocusChange(_) => {}
             TextEvent::Ime(e) => {
                 // TODO: Handle the cursor movement things from https://github.com/rust-windowing/winit/pull/3824
-                tracing::warn!(event = ?e, "TextArea doesn't accept IME");
+                let (fctx, lctx) = ctx.text_contexts();
+                // The text to submit as the TextChanged action. We do not send a TextChange action
+                // for the "virtual" preedit text.
+                let mut submit_text = None;
+                match e {
+                    winit::event::Ime::Disabled => {
+                        self.editor.transact(fctx, lctx, |txn| txn.clear_compose());
+                    }
+                    winit::event::Ime::Preedit(text, cursor) => {
+                        if text.is_empty() {
+                            self.editor.transact(fctx, lctx, |txn| txn.clear_compose());
+                        } else {
+                            if !self.editor.is_composing() && self.editor.selected_text().is_some()
+                            {
+                                // The IME has started composing. Delete the current selection and
+                                // send a TextChange event with the selection removed, but without
+                                // the composing preedit text.
+                                self.editor.transact(fctx, lctx, |txn| {
+                                    txn.delete_selection();
+                                });
+                                submit_text = Some(self.text().to_string());
+                            }
+
```

Comment:
Note `PlainEditorTxn::set_compose` also removes the selection from the buffer when the IME newly starts composing (i.e., the first compose after IME is enabled or after a `clear_compose`). That seemed like a useful abstraction for Parley. The Masonry code here as a special-case also removes the selection, because Masonry might not want to send preedit text in TextChange actions, but does want to notify about the removed selection.